src/BPEssentials/ExtensionMethods/PlayerWarns.cs

```diff
@@ -13,45 +15,103 @@ public class SerializableWarn
         {
             public string Reason { get; set; }
 
+            public bool Expired { get; set; }
+
             public string IssueraccountID { get; set; }
 
-            public DateTime Date { get; set; }
+            public DateTimeOffset Date { get; set; }
+
+            public int Length { get; set; }
 
-            public SerializableWarn(string issuer, string reason, DateTime dateTime)
+            public SerializableWarn(string issuer, string reason, DateTimeOffset dateTime, int length = -1)
             {
                 IssueraccountID = issuer;
                 Reason = reason;
                 Date = dateTime;
+                if (length < 1)
+                {
+                    Length = length;
+                }
+                else
+                {
+                    Length = Core.Instance.Settings.Warns.DefaultWarnsExpirationInDays;
+                }
             }
 
             public string ToString(ShPlayer player)
             {
                 var issuer = Core.Instance.SvManager.database.Users.FindById(IssueraccountID);
-                return player.T("warn_toString", Reason, issuer != null ? issuer.ID : IssueraccountID, Date.ToString(CultureInfo.InvariantCulture));
+                return player.T("warn_toString", Reason, issuer != null ? issuer.ID : IssueraccountID, Date.ToString(CultureInfo.InvariantCulture), Length, Expired ? player.T("warn_expired") : "");
```

Comment:
You can use a null coalescing operator here for a more clear readability in replacement of the ternary operator.
```suggestion
                return player.T("warn_toString", Reason, issuer?.ID ?? IssueraccountID, Date.ToString(CultureInfo.InvariantCulture), Length, Expired ? player.T("warn_expired") : "");
```